File: flink-scala-project/src/main/scala/kaichi/flink/exercises/TaxiRideCleansing.scala
package kaichi.flink.exercises

import com.dataartisans.flinktraining.exercises.datastream_java.sources.TaxiRideSource
import com.dataartisans.flinktraining.exercises.datastream_java.utils.GeoUtils
import org.apache.flink.api.java.utils.ParameterTool
import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.scala._

object TaxiRideCleansing {
  def main(args: Array[String]): Unit = {
    val params = ParameterTool.fromArgs(args)
    val input = params.getRequired("input")

    val maxDelay = 60
    val servingSpeed = 6

    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)

    val rides = env.addSource(new TaxiRideSource(input, maxDelay, servingSpeed))

    val nycRides = rides
      .filter(x => GeoUtils.isInNYC(x.startLon, x.startLat) && GeoUtils.isInNYC(x.endLon, x.endLat))

    nycRides.print()

    env.execute("Taxi Ride NYC Filtering")
  }

}

File: flink-java-project/src/main/java/kaichi/flink/exercises/TaxiRideCleansing.java
package kaichi.flink.exercises;

import com.dataartisans.flinktraining.exercises.datastream_java.datatypes.TaxiRide;
import com.dataartisans.flinktraining.exercises.datastream_java.sources.TaxiRideSource;
import com.dataartisans.flinktraining.exercises.datastream_java.utils.GeoUtils;
import org.apache.flink.api.common.functions.FilterFunction;
import org.apache.flink.api.java.utils.ParameterTool;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

public class TaxiRideCleansing {
    public static void main(String[] args) throws Exception {

        ParameterTool params = ParameterTool.fromArgs(args);
        final String input = params.getRequired("input");
        final int maxDelay = 60;
        final int maxSpeed = 100;

        StreamExecutionEnvironment env =
                StreamExecutionEnvironment.getExecutionEnvironment();
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);

        DataStream<TaxiRide> rides = env.addSource(
                new TaxiRideSource(input, maxDelay, maxSpeed));

        DataStream<TaxiRide> filteredRides = rides
                .filter(new NYCFilter());


        filteredRides.print();

        env.execute("Taxi Ride Cleansing");
    }

    public static class NYCFilter implements FilterFunction<TaxiRide> {

        @Override
        public boolean filter(TaxiRide taxiRide) {
            return GeoUtils.isInNYC(taxiRide.startLon, taxiRide.startLat) &&
                    GeoUtils.isInNYC(taxiRide.endLon, taxiRide.endLat);
        }
    }
}
